zjit/src/cruby_methods.rs

```diff
@@ -290,3 +291,111 @@ fn inline_basic_object_initialize(fun: &mut hir::Function, block: hir::BlockId,
     let result = fun.push_insn(block, hir::Insn::Const { val: hir::Const::Value(Qnil) });
     Some(result)
 }
+
+fn inline_kernel_respond_to_p(
+    fun: &mut hir::Function,
+    block: hir::BlockId,
+    recv: hir::InsnId,
+    args: &[hir::InsnId],
+    state: hir::InsnId,
+) -> Option<hir::InsnId> {
+    // Parse arguments: respond_to?(method_name, allow_priv = false)
+    let (method_name, allow_priv) = match *args {
+        [method_name] => (method_name, false),
+        [method_name, arg] => match fun.type_of(arg) {
+            t if t.is_known_truthy() => (method_name, true),
+            t if t.is_known_falsy() => (method_name, false),
+            _ => return None,
+        },
+        _ => return None,
+    };
+
+    // Method name must be a static symbol
+
+    let method_name = fun.type_of(method_name).ruby_object()?;
+    if !method_name.static_sym_p() {
+        return None;
+    }
+
+    // The receiver must have a known class to call respond_to? on
+    let recv_class = fun.type_of(recv).runtime_exact_ruby_class()?;
+
```

Comment:
```suggestion
    // The receiver must have a known class to call respond_to? on
    // TODO: This is technically overly strict. This would also work if all of the observed objects at this point agree on `respond_to?` and we can add many patchpoints.
    let recv_class = fun.type_of(recv).runtime_exact_ruby_class()?;

```